<!DOCTYPE html>
<html lang="en"><head>
<script src="DESeq2-Visualisation_files/libs/clipboard/clipboard.min.js"></script>
<script src="DESeq2-Visualisation_files/libs/quarto-html/tabby.min.js"></script>
<script src="DESeq2-Visualisation_files/libs/quarto-html/popper.min.js"></script>
<script src="DESeq2-Visualisation_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="DESeq2-Visualisation_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="DESeq2-Visualisation_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="DESeq2-Visualisation_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.2.269">

  <title>DESeq2-Visualisation</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="DESeq2-Visualisation_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="DESeq2-Visualisation_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1.6em;
      vertical-align: middle;
    }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="DESeq2-Visualisation_files/libs/revealjs/dist/theme/quarto.css" id="theme">
  <link href="DESeq2-Visualisation_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="DESeq2-Visualisation_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="DESeq2-Visualisation_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="DESeq2-Visualisation_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-captioned.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-captioned) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-captioned.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-captioned .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-captioned .callout-caption  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-captioned.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-captioned.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-caption {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-caption {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-caption {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-captioned .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-captioned .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-captioned) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-caption {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-caption {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-caption {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-caption {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-caption {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  
  <script src="DESeq2-Visualisation_files/libs/kePrint-0.0.1/kePrint.js"></script>
  <link href="DESeq2-Visualisation_files/libs/lightable-0.0.1/lightable.css" rel="stylesheet">
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">DESeq2-Visualisation</h1>

<div class="quarto-title-authors">
</div>

</section>
<section id="overview" class="slide level2">
<h2>Overview</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>Loading packages<br>
</li>
<li>Loading saved data<br>
</li>
<li>Dispersion plots<br>
</li>
<li>PCA plots<br>
</li>
<li>Heatmap</li>
</ul>
</div><div class="column" style="width:50%;">
<ul>
<li>Pairwise comparison<br>
</li>
<li>MA plot<br>
</li>
<li>Volcano plot<br>
</li>
<li>Results table</li>
</ul>
</div>
</div>
</section>
<section id="cran-packages" class="slide level2 smaller">
<h2>CRAN packages</h2>
<p>Install CRAN packages</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>cran.packages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"tidyverse"</span>,</span>
<span id="cb1-2"><a href="#cb1-2"></a>                   <span class="st">"ggrepel"</span>,</span>
<span id="cb1-3"><a href="#cb1-3"></a>                   <span class="st">"kableExtra"</span>,</span>
<span id="cb1-4"><a href="#cb1-4"></a>                   <span class="st">"RColorBrewer"</span>,</span>
<span id="cb1-5"><a href="#cb1-5"></a>                   <span class="st">"pheatmap"</span>,</span>
<span id="cb1-6"><a href="#cb1-6"></a>                   <span class="st">"devtools"</span>)</span>
<span id="cb1-7"><a href="#cb1-7"></a></span>
<span id="cb1-8"><a href="#cb1-8"></a>cran.load <span class="ot">&lt;-</span> <span class="cf">function</span>(pkg){</span>
<span id="cb1-9"><a href="#cb1-9"></a>        new.pkg <span class="ot">&lt;-</span> pkg[<span class="sc">!</span>(pkg <span class="sc">%in%</span> <span class="fu">installed.packages</span>()[, <span class="st">"Package"</span>])]</span>
<span id="cb1-10"><a href="#cb1-10"></a>        <span class="cf">if</span> (<span class="fu">length</span>(new.pkg)){</span>
<span id="cb1-11"><a href="#cb1-11"></a>          <span class="fu">install.packages</span>(new.pkg, <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-12"><a href="#cb1-12"></a>          }</span>
<span id="cb1-13"><a href="#cb1-13"></a>        <span class="fu">sapply</span>(pkg, require, <span class="at">character.only =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-14"><a href="#cb1-14"></a>}</span>
<span id="cb1-15"><a href="#cb1-15"></a><span class="fu">cran.load</span>(cran.packages)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="bioconductor-packages" class="slide level2 smaller">
<h2>Bioconductor Packages</h2>
<p>Install Biocondustor packages</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>bioconductor.packages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"DESeq2"</span>,</span>
<span id="cb2-2"><a href="#cb2-2"></a>                           <span class="st">"biomaRt"</span>,</span>
<span id="cb2-3"><a href="#cb2-3"></a>                           <span class="st">"apeglm"</span>)</span>
<span id="cb2-4"><a href="#cb2-4"></a></span>
<span id="cb2-5"><a href="#cb2-5"></a>bioconductor.load <span class="ot">&lt;-</span> <span class="cf">function</span>(pkg){</span>
<span id="cb2-6"><a href="#cb2-6"></a>        new.pkg <span class="ot">&lt;-</span> pkg[<span class="sc">!</span>(pkg <span class="sc">%in%</span> <span class="fu">installed.packages</span>()[, <span class="st">"Package"</span>])]</span>
<span id="cb2-7"><a href="#cb2-7"></a>        <span class="cf">if</span> (<span class="fu">length</span>(new.pkg)){</span>
<span id="cb2-8"><a href="#cb2-8"></a>          <span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb2-9"><a href="#cb2-9"></a>            <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb2-10"><a href="#cb2-10"></a>            BiocManager<span class="sc">::</span><span class="fu">install</span>(new.pkg)</span>
<span id="cb2-11"><a href="#cb2-11"></a>        }</span>
<span id="cb2-12"><a href="#cb2-12"></a>        <span class="fu">sapply</span>(pkg, require, <span class="at">character.only =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-13"><a href="#cb2-13"></a>}</span>
<span id="cb2-14"><a href="#cb2-14"></a><span class="fu">bioconductor.load</span>(bioconductor.packages)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="github-packages" class="slide level2 smaller">
<h2>Github Packages</h2>
<p>Install package fro GitHub</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r code-overflow-scroll number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>git.packages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"pipelineTools"</span>)</span>
<span id="cb3-2"><a href="#cb3-2"></a>repository <span class="ot">&lt;-</span> <span class="st">"GrahamHamilton"</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>git.load <span class="ot">&lt;-</span> <span class="cf">function</span>(pkg){</span>
<span id="cb3-4"><a href="#cb3-4"></a>        new.pkg <span class="ot">&lt;-</span> pkg[<span class="sc">!</span>(pkg <span class="sc">%in%</span> <span class="fu">installed.packages</span>()[, <span class="st">"Package"</span>])]</span>
<span id="cb3-5"><a href="#cb3-5"></a>        <span class="cf">if</span> (<span class="fu">length</span>(new.pkg))</span>
<span id="cb3-6"><a href="#cb3-6"></a>          <span class="fu">install_github</span>(<span class="fu">paste</span>(repository,new.pkg,<span class="at">sep =</span> <span class="st">"/"</span>), <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-7"><a href="#cb3-7"></a>        <span class="fu">sapply</span>(pkg, require, <span class="at">character.only =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-8"><a href="#cb3-8"></a>}</span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="fu">git.load</span>(git.packages)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-save-data" class="slide level2">
<h2>Load Save Data</h2>
<p>Load the data from the previous R session</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="fu">load</span>(<span class="st">"../Session_3/Session3.RData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="results-directory" class="slide level2">
<h2>Results Directory</h2>
<p>Create a results directory for the DESeq2 results files</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>deseq2_results_dir <span class="ot">&lt;-</span> <span class="st">"DESeq2"</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="fu">dir.create</span>(deseq2_results_dir, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="dispersion" class="slide level2 smaller">
<h2>Dispersion</h2>
<p>Dispersion is a measure of the spread of the data. DESeq2 dispersion estimates are:</p>
<ul>
<li>inversely related to the mean</li>
<li>directly related to variance</li>
</ul>
<p>Based on this relationship, the dispersion is higher for small mean counts and lower for large mean counts.</p>
<p>Dispersion estimates reflect the variance in gene expression for a given mean value.</p>
</section>
<section id="dispersion-plots" class="slide level2 smaller">
<h2>Dispersion plots</h2>
<div class="columns">
<div class="column" style="width:40%;">
<ul>
<li>Black points - dispersion estimates for each gene</li>
<li>Red line - fitted to the gene-wise dispersion estimates</li>
<li>Blue points - gene-wise dispersion estimates moved toward the expected dispersion values, “shrinking”</li>
</ul>
</div><div class="column" style="width:60%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="DESeq2-Visualisation_files/figure-revealjs/dipersion%20plot-1.png" width="960"></p>
</div>
</div>
</div>
</div>
</section>
<section id="principle-component-analysis" class="slide level2 smaller">
<h2>Principle Component Analysis</h2>
<p>Principal Component Analysis (PCA) is a technique used to emphasize variation and bring out strong patterns in a dataset<sup>1</sup>.</p>
<p>Plot expression values of Sample 1 versus Sample 2 <img data-src="images/PCA_1.png" style="width:60.0%"></p>

<aside><div>
<p>A good PCA tutorial by Josh Starmer</p>
</div><ol class="aside-footnotes"><li id="fn1"><p>https://www.youtube.com/watch?v=_UVHneBUBW0</p></li></ol></aside></section>
<section id="pca" class="slide level2 smaller">
<h2>PCA</h2>
<ul>
<li>Draw line through the data with the most variation, PC1</li>
<li>Draw line through the data with the second most variation, PC2</li>
<li>PCA will compute a per sample score</li>
</ul>

<img data-src="images/PCA_2.png" style="width:100.0%" class="r-stretch"></section>
<section id="pca-1" class="slide level2 smaller">
<h2>PCA</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>Dataset have larger dimensions</li>
<li>n-dimensional space with n axes</li>
<li>Principle component matrix with samples and scores</li>
<li>Plot principal components against each other</li>
<li>Expect that biological replicates to have similar scores</li>
<li>Samples from different treatment groups have different scores</li>
</ul>
</div><div class="column" style="width:50%;">
<p><img data-src="images/PCA_3.png" style="width:50.0%"> <img data-src="images/PCA_4.png" style="width:50.0%"></p>
</div>
</div>
</section>
<section id="pca-2" class="slide level2 smaller">
<h2>PCA</h2>
<p>Example PCA data</p>

<img data-src="images/PCA_table.png" style="width:50.0%" class="r-stretch"></section>
<section id="pca-3" class="slide level2 smaller">
<h2>PCA</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div style="font-size: 80%;">
<p>Treatment - No variation on PC1 or PC2 <img data-src="images/PCA_treatmentPC1.png" style="width:80.0%"></p>
<p>Cages - No variation on PC1 or PC2 <img data-src="images/PCA_cage.png" style="width:80.0%"></p>
</div>
</div><div class="column" style="width:50%;">
<div style="font-size: 80%;">
<p>Strain - variation PC1 <img data-src="images/PCA_strain.png" style="width:80.0%"></p>
<p>Sex - variation PC2 <img data-src="images/PCA_sex.png" style="width:80.0%"></p>
</div>
</div>
</div>
</section>
<section id="pca-4" class="slide level2 smaller">
<h2>PCA</h2>
<p>Explore other Principle components for variation due to <strong>Treatment</strong></p>

<img data-src="images/PCA_treatment_PC3.png" style="width:100.0%" class="r-stretch"></section>
<section id="pca-5" class="slide level2">
<h2>PCA</h2>
<ul>
<li>Samples separate by treatment on PC3</li>
<li>Can regress out the variation driving PC1 and PC2
<ul>
<li>Add Strain and Sex as confounding factors to the model</li>
</ul></li>
</ul>
</section>
<section id="pca-6" class="slide level2">
<h2>PCA</h2>
<p>Count data transformations by regularized logarithm and variance stabilizing transformations, used for Principle Component Analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>rld <span class="ot">&lt;-</span> <span class="fu">rlog</span>(dds,<span class="at">fitType =</span> <span class="st">"local"</span>)</span>
<span id="cb6-2"><a href="#cb6-2"></a>rld_mat <span class="ot">&lt;-</span> <span class="fu">assay</span>(rld)</span>
<span id="cb6-3"><a href="#cb6-3"></a>vsd <span class="ot">&lt;-</span> <span class="fu">varianceStabilizingTransformation</span>(dds,<span class="at">fitType =</span> <span class="st">"local"</span>)</span>
<span id="cb6-4"><a href="#cb6-4"></a>vsd_mat <span class="ot">&lt;-</span> <span class="fu">assay</span>(vsd)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="pca-plots" class="slide level2 smaller">
<h2>PCA Plots</h2>
<p>Calculate the principle component values using the R package prcomp(). Add the meta data from the sample description table.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(rld_mat))</span>
<span id="cb7-2"><a href="#cb7-2"></a></span>
<span id="cb7-3"><a href="#cb7-3"></a>pca_data <span class="ot">&lt;-</span> pca<span class="sc">$</span>x <span class="sc">%&gt;%</span></span>
<span id="cb7-4"><a href="#cb7-4"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"Sample"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb7-6"><a href="#cb7-6"></a>  <span class="fu">left_join</span>(sample_table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<div style="font-size: 50%;">
<div class="cell">
<div class="cell-output-display">

<table class="table" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;"> Sample </th>
   <th style="text-align:right;"> PC1 </th>
   <th style="text-align:right;"> PC2 </th>
   <th style="text-align:right;"> PC3 </th>
   <th style="text-align:right;"> PC4 </th>
   <th style="text-align:right;"> PC5 </th>
   <th style="text-align:right;"> PC6 </th>
   <th style="text-align:right;"> PC7 </th>
   <th style="text-align:right;"> PC8 </th>
   <th style="text-align:left;"> Condition </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> Control_1 </td>
   <td style="text-align:right;"> -86.28054 </td>
   <td style="text-align:right;"> -34.041901 </td>
   <td style="text-align:right;"> 19.1629679 </td>
   <td style="text-align:right;"> -1.0008706 </td>
   <td style="text-align:right;"> 0.2727963 </td>
   <td style="text-align:right;"> -0.1937691 </td>
   <td style="text-align:right;"> -0.2299436 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> Control </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Control_2 </td>
   <td style="text-align:right;"> -87.42798 </td>
   <td style="text-align:right;"> -35.173093 </td>
   <td style="text-align:right;"> -18.8606179 </td>
   <td style="text-align:right;"> 0.8237575 </td>
   <td style="text-align:right;"> -0.1618021 </td>
   <td style="text-align:right;"> 0.0756380 </td>
   <td style="text-align:right;"> 0.2628584 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> Control </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Control_3 </td>
   <td style="text-align:right;"> -48.82317 </td>
   <td style="text-align:right;"> 47.335772 </td>
   <td style="text-align:right;"> 0.8659648 </td>
   <td style="text-align:right;"> 15.3859036 </td>
   <td style="text-align:right;"> 2.3232593 </td>
   <td style="text-align:right;"> -0.5460535 </td>
   <td style="text-align:right;"> -0.9021464 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> Control </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Control_4 </td>
   <td style="text-align:right;"> -44.71270 </td>
   <td style="text-align:right;"> 46.267735 </td>
   <td style="text-align:right;"> -1.1147754 </td>
   <td style="text-align:right;"> -15.6747137 </td>
   <td style="text-align:right;"> -2.8408276 </td>
   <td style="text-align:right;"> 0.7682248 </td>
   <td style="text-align:right;"> 0.7150646 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> Control </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Treatment_1 </td>
   <td style="text-align:right;"> 65.74631 </td>
   <td style="text-align:right;"> -4.817170 </td>
   <td style="text-align:right;"> -0.7033761 </td>
   <td style="text-align:right;"> -3.2193434 </td>
   <td style="text-align:right;"> 13.3075631 </td>
   <td style="text-align:right;"> -10.1391418 </td>
   <td style="text-align:right;"> -1.3050256 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> Treatment </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Treatment_2 </td>
   <td style="text-align:right;"> 68.01609 </td>
   <td style="text-align:right;"> -7.133756 </td>
   <td style="text-align:right;"> -0.2467674 </td>
   <td style="text-align:right;"> 0.1463250 </td>
   <td style="text-align:right;"> -3.7462348 </td>
   <td style="text-align:right;"> 6.1030450 </td>
   <td style="text-align:right;"> -13.7742179 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> Treatment </td>
  </tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="pca-plots-1" class="slide level2 smaller">
<h2>PCA Plots</h2>
<p>Can also calculate the percentage variaction for each principle component</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>percent_var <span class="ot">&lt;-</span> <span class="fu">round</span>(pca<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span><span class="sc">/</span><span class="fu">sum</span>(pca<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span>) <span class="sc">*</span> <span class="dv">100</span>, <span class="at">digits =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">

<table class="table table-condensed" style="margin-left: auto; margin-right: auto;">
<tbody>
  <tr>
   <td style="text-align:right;"> 80.2 </td>
  </tr>
  <tr>
   <td style="text-align:right;"> 14.9 </td>
  </tr>
  <tr>
   <td style="text-align:right;"> 1.6 </td>
  </tr>
  <tr>
   <td style="text-align:right;"> 1.1 </td>
  </tr>
  <tr>
   <td style="text-align:right;"> 0.9 </td>
  </tr>
  <tr>
   <td style="text-align:right;"> 0.7 </td>
  </tr>
</tbody>
</table>

</div>
</div>
</section>
<section id="pca-plots-2" class="slide level2 smaller">
<h2>PCA Plots</h2>
<p>Plot PC1 against PC2 using ggplot2</p>

<img data-src="DESeq2-Visualisation_files/figure-revealjs/unnamed-chunk-7-1.png" width="960" class="r-stretch"></section>
<section id="pca-plots-3" class="slide level2 smaller">
<h2>PCA Plots</h2>
<p>Similar plot using the vsd expression values</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(vsd_mat))</span>
<span id="cb9-2"><a href="#cb9-2"></a>pca_data <span class="ot">&lt;-</span> pca<span class="sc">$</span>x <span class="sc">%&gt;%</span></span>
<span id="cb9-3"><a href="#cb9-3"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb9-4"><a href="#cb9-4"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"Sample"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb9-5"><a href="#cb9-5"></a>  <span class="fu">left_join</span>(sample_table)</span>
<span id="cb9-6"><a href="#cb9-6"></a></span>
<span id="cb9-7"><a href="#cb9-7"></a>percent_var <span class="ot">&lt;-</span> <span class="fu">round</span>(pca<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span><span class="sc">/</span><span class="fu">sum</span>(pca<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span>) <span class="sc">*</span> <span class="dv">100</span>, <span class="at">digits =</span> <span class="dv">1</span>)</span>
<span id="cb9-8"><a href="#cb9-8"></a></span>
<span id="cb9-9"><a href="#cb9-9"></a>vsd_pca_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pca_data,<span class="fu">aes</span>(PC1,PC2, <span class="at">label=</span>Sample, <span class="at">color=</span>Condition)) <span class="sc">+</span></span>
<span id="cb9-10"><a href="#cb9-10"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb9-11"><a href="#cb9-11"></a>  <span class="fu">geom_text_repel</span>(<span class="at">colour =</span> <span class="st">"black"</span>,<span class="at">size=</span><span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb9-12"><a href="#cb9-12"></a>  <span class="fu">xlab</span>(<span class="fu">paste0</span>(<span class="st">"PC1: "</span>,percent_var[<span class="dv">1</span>],<span class="st">"% variance"</span>)) <span class="sc">+</span></span>
<span id="cb9-13"><a href="#cb9-13"></a>  <span class="fu">ylab</span>(<span class="fu">paste0</span>(<span class="st">"PC2: "</span>,percent_var[<span class="dv">2</span>],<span class="st">"% variance"</span>)) <span class="sc">+</span></span>
<span id="cb9-14"><a href="#cb9-14"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb9-15"><a href="#cb9-15"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"PCA Plot"</span>,</span>
<span id="cb9-16"><a href="#cb9-16"></a>       <span class="at">subtitle =</span> <span class="st">"vsd transformed "</span>) <span class="sc">+</span></span>
<span id="cb9-17"><a href="#cb9-17"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>,</span>
<span id="cb9-18"><a href="#cb9-18"></a>        <span class="at">panel.grid.major.x =</span> <span class="fu">element_line</span>(<span class="st">"grey"</span>), </span>
<span id="cb9-19"><a href="#cb9-19"></a>        <span class="at">panel.grid.major.y=</span> <span class="fu">element_line</span>(<span class="st">"grey"</span>))</span>
<span id="cb9-20"><a href="#cb9-20"></a></span>
<span id="cb9-21"><a href="#cb9-21"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">"vsd_PCA.png"</span>,</span>
<span id="cb9-22"><a href="#cb9-22"></a>       <span class="at">plot =</span> vsd_pca_plot,</span>
<span id="cb9-23"><a href="#cb9-23"></a>       <span class="at">path =</span> deseq2_results_dir)</span>
<span id="cb9-24"><a href="#cb9-24"></a></span>
<span id="cb9-25"><a href="#cb9-25"></a>vsd_pca_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

</section>
<section id="pca-plots-3-output" class="slide level2 smaller output-location-slide"><h2>PCA Plots</h2><div class="cell output-location-slide">
<div class="cell-output-display">
<p><img data-src="DESeq2-Visualisation_files/figure-revealjs/vsd%20PCA-1.png" width="960"></p>
</div>
</div></section><section id="heatmap" class="slide level2 smaller">
<h2>Heatmap</h2>
<p>Get a list ENSEMBL gene ids, with a p adjusted value of 0.1 or less</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>de_genes <span class="ot">&lt;-</span> <span class="fu">results</span>(dds) <span class="sc">%&gt;%</span> <span class="co"># Get the de results</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> <span class="co"># Convert results to a data frame</span></span>
<span id="cb10-3"><a href="#cb10-3"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"ensembl_gene_id"</span>) <span class="sc">%&gt;%</span> <span class="co"># add the ensembl ids as a data frame column</span></span>
<span id="cb10-4"><a href="#cb10-4"></a>  <span class="fu">filter</span>(padj <span class="sc">&lt;=</span> qvalueCutoff) <span class="sc">%&gt;%</span> <span class="co"># Select data frame rows that have p values below the cut off</span></span>
<span id="cb10-5"><a href="#cb10-5"></a>  <span class="fu">pull</span>(ensembl_gene_id) <span class="co"># get the ensembl ids of the filtered genes as a character vector</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<div class="cell">
<div class="cell-output-display">

<table class="table" style="margin-left: auto; margin-right: auto;">
<tbody>
  <tr>
   <td style="text-align:left;"> ENSG00000188976 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000187961 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000187583 </td>
  </tr>
</tbody>
</table>

</div>
</div>
</section>
<section id="heatmap-1" class="slide level2 smaller">
<h2>Heatmap</h2>
<p>Create a table of log transformed normalised read counts. Convert into Z-scores.</p>
<p>Z-scores - Take the log of the read counts, subtract the mean for the gene then divide by the standard deviation for that gene.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r code-overflow-wrap number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>log2_de_genes <span class="ot">&lt;-</span> total_norm_counts  <span class="sc">%&gt;%</span></span>
<span id="cb11-2"><a href="#cb11-2"></a>  <span class="fu">filter</span>(ensembl_gene_id <span class="sc">%in%</span> de_genes) <span class="sc">%&gt;%</span> <span class="co"># Filter the normalised counts data frame for rows with the ensembl id of the differentially expressed genes list</span></span>
<span id="cb11-3"><a href="#cb11-3"></a>  <span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">"ensembl_gene_id"</span>) <span class="sc">%&gt;%</span> <span class="co"># convert the ensembl_gene_id column to the dtaframe rownames</span></span>
<span id="cb11-4"><a href="#cb11-4"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">where</span>(is.numeric), <span class="sc">~</span> .x <span class="sc">+</span> <span class="dv">1</span>)) <span class="sc">%&gt;%</span> <span class="co"># Add 1 to all the values to avoid NaN errors</span></span>
<span id="cb11-5"><a href="#cb11-5"></a>  <span class="fu">log2</span>() <span class="co"># Log2 tranform the data frame values</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 50%;">
<div class="cell">
<div class="cell-output-display">

<table class="table table-condensed" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;">   </th>
   <th style="text-align:right;"> Control_1 </th>
   <th style="text-align:right;"> Control_2 </th>
   <th style="text-align:right;"> Control_3 </th>
   <th style="text-align:right;"> Control_4 </th>
   <th style="text-align:right;"> Treatment_1 </th>
   <th style="text-align:right;"> Treatment_2 </th>
   <th style="text-align:right;"> Treatment_3 </th>
   <th style="text-align:right;"> Treatment_4 </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> ENSG00000188976 </td>
   <td style="text-align:right;"> 9.141637 </td>
   <td style="text-align:right;"> 9.251128 </td>
   <td style="text-align:right;"> 9.647133 </td>
   <td style="text-align:right;"> 9.635403 </td>
   <td style="text-align:right;"> 10.03776 </td>
   <td style="text-align:right;"> 9.928693 </td>
   <td style="text-align:right;"> 9.924111 </td>
   <td style="text-align:right;"> 10.11613 </td>
  </tr>
</tbody>
</table>

</div>
</div>
<p><br></p>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r code-overflow-wrap number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>z_scores <span class="ot">&lt;-</span> (log2_de_genes <span class="sc">-</span> <span class="fu">rowMeans</span>(log2_de_genes)) <span class="sc">/</span><span class="fu">rowSds</span>(<span class="fu">as.matrix</span>(log2_de_genes)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 50%;">
<div class="cell">
<div class="cell-output-display">

<table class="table table-condensed" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;">   </th>
   <th style="text-align:right;"> Control_1 </th>
   <th style="text-align:right;"> Control_2 </th>
   <th style="text-align:right;"> Control_3 </th>
   <th style="text-align:right;"> Control_4 </th>
   <th style="text-align:right;"> Treatment_1 </th>
   <th style="text-align:right;"> Treatment_2 </th>
   <th style="text-align:right;"> Treatment_3 </th>
   <th style="text-align:right;"> Treatment_4 </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> ENSG00000188976 </td>
   <td style="text-align:right;"> -1.577782 </td>
   <td style="text-align:right;"> -1.273966 </td>
   <td style="text-align:right;"> -0.1751317 </td>
   <td style="text-align:right;"> -0.2076823 </td>
   <td style="text-align:right;"> 0.9087651 </td>
   <td style="text-align:right;"> 0.6061395 </td>
   <td style="text-align:right;"> 0.593426 </td>
   <td style="text-align:right;"> 1.126232 </td>
  </tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="heatmap-2" class="slide level2 smaller">
<h2>Heatmap</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>num_colours <span class="ot">=</span> <span class="dv">100</span> </span>
<span id="cb13-2"><a href="#cb13-2"></a></span>
<span id="cb13-3"><a href="#cb13-3"></a>symetric_breaks <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="fu">max</span>(<span class="fu">abs</span>(z_scores)),</span>
<span id="cb13-4"><a href="#cb13-4"></a>                       <span class="fu">max</span>(<span class="fu">abs</span>(z_scores)),</span>
<span id="cb13-5"><a href="#cb13-5"></a>                       <span class="at">length.out =</span> num_colours <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb13-6"><a href="#cb13-6"></a></span>
<span id="cb13-7"><a href="#cb13-7"></a>column_annotations <span class="ot">&lt;-</span> sample_table <span class="sc">%&gt;%</span></span>
<span id="cb13-8"><a href="#cb13-8"></a>  <span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">"Sample"</span>)</span>
<span id="cb13-9"><a href="#cb13-9"></a></span>
<span id="cb13-10"><a href="#cb13-10"></a>ph <span class="ot">&lt;-</span> <span class="fu">pheatmap</span>(z_scores,</span>
<span id="cb13-11"><a href="#cb13-11"></a>         <span class="at">color =</span><span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"#003865"</span>, <span class="st">"#FFFFFF"</span>,<span class="st">"#B30C00"</span>))(num_colours),</span>
<span id="cb13-12"><a href="#cb13-12"></a>         <span class="at">breaks =</span> symetric_breaks,</span>
<span id="cb13-13"><a href="#cb13-13"></a>         <span class="at">show_rownames =</span> F,</span>
<span id="cb13-14"><a href="#cb13-14"></a>         <span class="at">annotation_col =</span> column_annotations)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>png 
  2 </code></pre>
</div>
</div>
</section>
<section id="heatmap-2-output" class="slide level2 smaller output-location-slide"><h2>Heatmap</h2><div class="cell output-location-slide">
<div class="cell-output-display">
<p><img data-src="DESeq2-Visualisation_files/figure-revealjs/unnamed-chunk-13-1.png" width="960"></p>
</div>
</div></section><section id="annotations" class="slide level2 smaller">
<h2>Annotations</h2>
<p>Create an annotations table using data from ENSEMBL Biomart</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15" data-code-line-numbers="1-10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>database <span class="ot">&lt;-</span> <span class="st">"hsapiens_gene_ensembl"</span></span>
<span id="cb15-2"><a href="#cb15-2"></a>mart <span class="ot">&lt;-</span> <span class="st">"genes"</span></span>
<span id="cb15-3"><a href="#cb15-3"></a>filt <span class="ot">&lt;-</span> <span class="st">"ensembl_gene_id"</span></span>
<span id="cb15-4"><a href="#cb15-4"></a></span>
<span id="cb15-5"><a href="#cb15-5"></a>ensembl <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="at">biomart =</span> mart, <span class="at">dataset =</span> database)</span>
<span id="cb15-6"><a href="#cb15-6"></a></span>
<span id="cb15-7"><a href="#cb15-7"></a>att <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ensembl_gene_id"</span>,<span class="st">"external_gene_name"</span>,<span class="st">"chromosome_name"</span>,<span class="st">"start_position"</span>,<span class="st">"end_position"</span>,<span class="st">"gene_biotype"</span>,<span class="st">"entrezgene_id"</span>)</span>
<span id="cb15-8"><a href="#cb15-8"></a></span>
<span id="cb15-9"><a href="#cb15-9"></a>annotations <span class="ot">&lt;-</span> <span class="fu">getBM</span>(<span class="at">attributes=</span>att,<span class="at">filter=</span>filt,<span class="at">values=</span><span class="fu">rownames</span>(dds),<span class="at">mart=</span>ensembl) <span class="sc">%&gt;%</span> </span>
<span id="cb15-10"><a href="#cb15-10"></a>  <span class="fu">distinct</span>(ensembl_gene_id, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">

</div>
<div class="cell">

</div>
<p><br></p>
<div style="font-size: 50%;">
<div class="cell">
<div class="cell-output-display">

<table class="table" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;"> ensembl_gene_id </th>
   <th style="text-align:left;"> external_gene_name </th>
   <th style="text-align:left;"> chromosome_name </th>
   <th style="text-align:right;"> start_position </th>
   <th style="text-align:right;"> end_position </th>
   <th style="text-align:left;"> gene_biotype </th>
   <th style="text-align:right;"> entrezgene_id </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> ENSG00000000457 </td>
   <td style="text-align:left;"> SCYL3 </td>
   <td style="text-align:left;"> 1 </td>
   <td style="text-align:right;"> 169849631 </td>
   <td style="text-align:right;"> 169894267 </td>
   <td style="text-align:left;"> protein_coding </td>
   <td style="text-align:right;"> 57147 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000000460 </td>
   <td style="text-align:left;"> C1orf112 </td>
   <td style="text-align:left;"> 1 </td>
   <td style="text-align:right;"> 169662007 </td>
   <td style="text-align:right;"> 169854080 </td>
   <td style="text-align:left;"> protein_coding </td>
   <td style="text-align:right;"> 55732 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000000938 </td>
   <td style="text-align:left;"> FGR </td>
   <td style="text-align:left;"> 1 </td>
   <td style="text-align:right;"> 27612064 </td>
   <td style="text-align:right;"> 27635185 </td>
   <td style="text-align:left;"> protein_coding </td>
   <td style="text-align:right;"> 2268 </td>
  </tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="pairwise-differential-expression" class="slide level2 smaller">
<h2>Pairwise differential expression</h2>
<p>Select the groups for comparison</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>base_condition <span class="ot">&lt;-</span> comparisons<span class="sc">$</span>Base_condition <span class="sc">%&gt;%</span> <span class="fu">as.character</span>() </span>
<span id="cb16-2"><a href="#cb16-2"></a>test_condition <span class="ot">&lt;-</span> comparisons<span class="sc">$</span>Test_condition <span class="sc">%&gt;%</span> <span class="fu">as.character</span>()</span>
<span id="cb16-3"><a href="#cb16-3"></a></span>
<span id="cb16-4"><a href="#cb16-4"></a><span class="co"># Contrast name</span></span>
<span id="cb16-5"><a href="#cb16-5"></a>contrast_name <span class="ot">&lt;-</span> <span class="fu">paste</span>(test_condition, <span class="st">"vs"</span>, base_condition, <span class="at">sep =</span> <span class="st">"_"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="pairwise-results" class="slide level2">
<h2>Pairwise Results</h2>
<p>Create directory for DESeq2 data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>results_dir <span class="ot">=</span> <span class="fu">paste</span>(contrast_name, <span class="st">"results"</span>, <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb17-2"><a href="#cb17-2"></a>resultsFilePath <span class="ot">&lt;-</span><span class="fu">file.path</span>(deseq2_results_dir, results_dir, <span class="at">fsep =</span> .Platform<span class="sc">$</span>file.sep)</span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="fu">dir.create</span>(resultsFilePath, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Get the sample names for the contrast</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>comparisonSampleNames <span class="ot">&lt;-</span>  sample_table <span class="sc">%&gt;%</span></span>
<span id="cb18-2"><a href="#cb18-2"></a>  <span class="fu">filter</span>(Condition <span class="sc">%in%</span> <span class="fu">c</span>(test_condition, base_condition)) <span class="sc">%&gt;%</span></span>
<span id="cb18-3"><a href="#cb18-3"></a>  <span class="fu">pull</span>(Sample)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="pairwise-results-1" class="slide level2 smaller">
<h2>Pairwise Results</h2>
<p>Get the log2 normalised counts to the results tables</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>contrast_total_norm_counts <span class="ot">&lt;-</span> total_norm_counts <span class="sc">%&gt;%</span> </span>
<span id="cb19-2"><a href="#cb19-2"></a>  <span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">"ensembl_gene_id"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb19-3"><a href="#cb19-3"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(comparisonSampleNames) <span class="sc">%&gt;%</span></span>
<span id="cb19-4"><a href="#cb19-4"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"ensembl_gene_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<div style="font-size: 40%;">
<div class="cell">
<div class="cell-output-display">

<table class="table" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;"> ensembl_gene_id </th>
   <th style="text-align:right;"> Control_1 </th>
   <th style="text-align:right;"> Control_2 </th>
   <th style="text-align:right;"> Control_3 </th>
   <th style="text-align:right;"> Control_4 </th>
   <th style="text-align:right;"> Treatment_1 </th>
   <th style="text-align:right;"> Treatment_2 </th>
   <th style="text-align:right;"> Treatment_3 </th>
   <th style="text-align:right;"> Treatment_4 </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> ENSG00000186092 </td>
   <td style="text-align:right;"> 0.00000 </td>
   <td style="text-align:right;"> 0.000000 </td>
   <td style="text-align:right;"> 0.00000 </td>
   <td style="text-align:right;"> 0.000000 </td>
   <td style="text-align:right;"> 0.00000 </td>
   <td style="text-align:right;"> 0.000000 </td>
   <td style="text-align:right;"> 0.000000 </td>
   <td style="text-align:right;"> 0.00000 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000284733 </td>
   <td style="text-align:right;"> 0.00000 </td>
   <td style="text-align:right;"> 0.000000 </td>
   <td style="text-align:right;"> 0.00000 </td>
   <td style="text-align:right;"> 0.000000 </td>
   <td style="text-align:right;"> 0.00000 </td>
   <td style="text-align:right;"> 0.000000 </td>
   <td style="text-align:right;"> 0.000000 </td>
   <td style="text-align:right;"> 0.00000 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000284662 </td>
   <td style="text-align:right;"> 0.00000 </td>
   <td style="text-align:right;"> 0.000000 </td>
   <td style="text-align:right;"> 0.00000 </td>
   <td style="text-align:right;"> 0.000000 </td>
   <td style="text-align:right;"> 0.00000 </td>
   <td style="text-align:right;"> 0.000000 </td>
   <td style="text-align:right;"> 0.000000 </td>
   <td style="text-align:right;"> 0.00000 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000187634 </td>
   <td style="text-align:right;"> 10.67159 </td>
   <td style="text-align:right;"> 3.755249 </td>
   <td style="text-align:right;"> 10.44547 </td>
   <td style="text-align:right;"> 2.743786 </td>
   <td style="text-align:right;"> 12.04994 </td>
   <td style="text-align:right;"> 7.806789 </td>
   <td style="text-align:right;"> 8.671382 </td>
   <td style="text-align:right;"> 9.70868 </td>
  </tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="pairwise-results-2" class="slide level2 smaller">
<h2>Pairwise Results</h2>
<p>Create the results table</p>
<ul>
<li><strong>The order of the names determines the direction of fold change reported</strong></li>
<li><strong>The second element is the level that is used as baseline</strong></li>
<li><strong>Observe a log2 fold change of -2 - gene expression is lower treatment relative to the control</strong></li>
</ul>
<p>Add annotations and normalised read counts to the results table using the dplyr left_join command</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast =</span> <span class="fu">c</span>(<span class="st">"Condition"</span>, test_condition, base_condition)) <span class="sc">%&gt;%</span> <span class="co"># Get the pair se comparison results</span></span>
<span id="cb20-2"><a href="#cb20-2"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> <span class="co"># Convert to dataframe</span></span>
<span id="cb20-3"><a href="#cb20-3"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"ensembl_gene_id"</span>) <span class="sc">%&gt;%</span> <span class="co"># Create an ensembl gene id column</span></span>
<span id="cb20-4"><a href="#cb20-4"></a>  <span class="fu">left_join</span>(annotations, <span class="at">by =</span> <span class="st">"ensembl_gene_id"</span>)<span class="sc">%&gt;%</span> <span class="co"># merge annotations by ensembl gene id</span></span>
<span id="cb20-5"><a href="#cb20-5"></a>  <span class="fu">left_join</span>(contrast_total_norm_counts, <span class="at">by =</span> <span class="st">"ensembl_gene_id"</span>) <span class="sc">%&gt;%</span> <span class="co"># Merge normalised counts by ensembl gene id</span></span>
<span id="cb20-6"><a href="#cb20-6"></a>  <span class="fu">arrange</span>(padj) <span class="sc">%&gt;%</span> <span class="co"># Sort the dataframe by the p adjusted value ascending</span></span>
<span id="cb20-7"><a href="#cb20-7"></a>  <span class="fu">replace_na</span>(<span class="fu">list</span>(<span class="at">log2FoldChange =</span> <span class="fl">0.001</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="pairwise-results-3" class="slide level2 smaller">
<h2>Pairwise Results</h2>
<ul>
<li>baseMean - mean of normalized counts for all samples<br>
</li>
<li>log2FoldChange - log2 fold change, condition treated vs untreated<br>
</li>
<li>lfcSE - standard error, condition treated vs untreated<br>
</li>
<li>stat - Wald statistic, condition treated vs untreated<br>
</li>
<li>pvalue - Wald test p-value: condition treated vs untreated<br>
</li>
<li>padj - BH adjusted p-values</li>
</ul>
<div style="font-size: 40%;">
<div class="cell">
<div class="cell-output-display">

<table class="table" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;"> ensembl_gene_id </th>
   <th style="text-align:right;"> baseMean </th>
   <th style="text-align:right;"> log2FoldChange </th>
   <th style="text-align:right;"> lfcSE </th>
   <th style="text-align:right;"> stat </th>
   <th style="text-align:right;"> pvalue </th>
   <th style="text-align:right;"> padj </th>
   <th style="text-align:left;"> external_gene_name </th>
   <th style="text-align:left;"> chromosome_name </th>
   <th style="text-align:right;"> start_position </th>
   <th style="text-align:right;"> end_position </th>
   <th style="text-align:left;"> gene_biotype </th>
   <th style="text-align:right;"> entrezgene_id </th>
   <th style="text-align:right;"> Control_1 </th>
   <th style="text-align:right;"> Control_2 </th>
   <th style="text-align:right;"> Control_3 </th>
   <th style="text-align:right;"> Control_4 </th>
   <th style="text-align:right;"> Treatment_1 </th>
   <th style="text-align:right;"> Treatment_2 </th>
   <th style="text-align:right;"> Treatment_3 </th>
   <th style="text-align:right;"> Treatment_4 </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> ENSG00000157191 </td>
   <td style="text-align:right;"> 8171.718 </td>
   <td style="text-align:right;"> -2.429669 </td>
   <td style="text-align:right;"> 0.0374863 </td>
   <td style="text-align:right;"> -64.81491 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> NECAP2 </td>
   <td style="text-align:left;"> 1 </td>
   <td style="text-align:right;"> 16440721 </td>
   <td style="text-align:right;"> 16460078 </td>
   <td style="text-align:left;"> protein_coding </td>
   <td style="text-align:right;"> 55707 </td>
   <td style="text-align:right;"> 13556.4812 </td>
   <td style="text-align:right;"> 13858.7451 </td>
   <td style="text-align:right;"> 14103.1250 </td>
   <td style="text-align:right;"> 13621.5255 </td>
   <td style="text-align:right;"> 2555.7914 </td>
   <td style="text-align:right;"> 2514.9013 </td>
   <td style="text-align:right;"> 2576.0675 </td>
   <td style="text-align:right;"> 2587.1079 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000117090 </td>
   <td style="text-align:right;"> 2322.100 </td>
   <td style="text-align:right;"> -5.224265 </td>
   <td style="text-align:right;"> 0.0775600 </td>
   <td style="text-align:right;"> -67.35771 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> SLAMF1 </td>
   <td style="text-align:left;"> 1 </td>
   <td style="text-align:right;"> 160608106 </td>
   <td style="text-align:right;"> 160647044 </td>
   <td style="text-align:left;"> protein_coding </td>
   <td style="text-align:right;"> 6504 </td>
   <td style="text-align:right;"> 4554.9919 </td>
   <td style="text-align:right;"> 4369.2318 </td>
   <td style="text-align:right;"> 4578.5976 </td>
   <td style="text-align:right;"> 4588.9821 </td>
   <td style="text-align:right;"> 118.6919 </td>
   <td style="text-align:right;"> 125.4663 </td>
   <td style="text-align:right;"> 127.4026 </td>
   <td style="text-align:right;"> 113.4383 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000026751 </td>
   <td style="text-align:right;"> 4495.926 </td>
   <td style="text-align:right;"> -4.690935 </td>
   <td style="text-align:right;"> 0.0870932 </td>
   <td style="text-align:right;"> -53.86110 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> SLAMF7 </td>
   <td style="text-align:left;"> 1 </td>
   <td style="text-align:right;"> 160739057 </td>
   <td style="text-align:right;"> 160754821 </td>
   <td style="text-align:left;"> protein_coding </td>
   <td style="text-align:right;"> 57823 </td>
   <td style="text-align:right;"> 8364.7508 </td>
   <td style="text-align:right;"> 8424.9003 </td>
   <td style="text-align:right;"> 9178.0860 </td>
   <td style="text-align:right;"> 8659.3886 </td>
   <td style="text-align:right;"> 366.9206 </td>
   <td style="text-align:right;"> 301.1190 </td>
   <td style="text-align:right;"> 320.1741 </td>
   <td style="text-align:right;"> 352.0674 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000072694 </td>
   <td style="text-align:right;"> 2202.988 </td>
   <td style="text-align:right;"> 4.309099 </td>
   <td style="text-align:right;"> 0.1114264 </td>
   <td style="text-align:right;"> 38.67216 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> FCGR2B </td>
   <td style="text-align:left;"> 1 </td>
   <td style="text-align:right;"> 161663143 </td>
   <td style="text-align:right;"> 161678654 </td>
   <td style="text-align:left;"> protein_coding </td>
   <td style="text-align:right;"> 2213 </td>
   <td style="text-align:right;"> 243.6681 </td>
   <td style="text-align:right;"> 229.0702 </td>
   <td style="text-align:right;"> 186.2775 </td>
   <td style="text-align:right;"> 190.6931 </td>
   <td style="text-align:right;"> 4103.6057 </td>
   <td style="text-align:right;"> 4273.6594 </td>
   <td style="text-align:right;"> 4433.7444 </td>
   <td style="text-align:right;"> 3963.1856 </td>
  </tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="pairwise-results-4" class="slide level2 smaller">
<h2>Pairwise Results</h2>
<p>Save the results table to a file</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="co"># Write the results data to file</span></span>
<span id="cb21-2"><a href="#cb21-2"></a><span class="co"># Output file name</span></span>
<span id="cb21-3"><a href="#cb21-3"></a>outfileName <span class="ot">&lt;-</span> <span class="fu">paste</span>(contrast_name, <span class="st">'genes'</span>, <span class="st">'tsv'</span>, <span class="at">sep =</span> <span class="st">"."</span>)</span>
<span id="cb21-4"><a href="#cb21-4"></a><span class="fu">write_delim</span>(contrast,</span>
<span id="cb21-5"><a href="#cb21-5"></a>            <span class="at">file =</span> <span class="fu">file.path</span>(resultsFilePath, outfileName, <span class="at">fsep =</span> .Platform<span class="sc">$</span>file.sep),</span>
<span id="cb21-6"><a href="#cb21-6"></a>            <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ma-plot" class="slide level2 smaller">
<h2>MA Plot</h2>
<ul>
<li>MA plot shows the mean of the normalized counts versus the log2 foldchanges.</li>
<li>Significantly differentially expressed genes are colored.</li>
</ul>
</section>
<section id="ma-plot-1" class="slide level2 smaller">
<h2>MA Plot</h2>
<p>Plot the log of the baseMean against the Log fold change</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>ma_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(contrast, <span class="fu">aes</span>(  <span class="at">x =</span> <span class="fu">log10</span>(baseMean), <span class="at">y =</span> log2FoldChange, <span class="at">colour =</span> padj <span class="sc">&lt;</span> <span class="fl">0.1</span>  )) <span class="sc">+</span></span>
<span id="cb22-2"><a href="#cb22-2"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb22-3"><a href="#cb22-3"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">name =</span> <span class="st">'padj &lt; 0.1'</span>, <span class="at">values =</span> <span class="fu">setNames</span>(<span class="fu">c</span>(<span class="st">"#003865"</span>, <span class="st">"#cccccc"</span>), <span class="fu">c</span>(T, F))) <span class="sc">+</span></span>
<span id="cb22-4"><a href="#cb22-4"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="fu">floor</span>(<span class="fu">min</span>(contrast<span class="sc">$</span>log2FoldChange)), <span class="fu">ceiling</span>(<span class="fu">max</span>(contrast<span class="sc">$</span>log2FoldChange)), )) <span class="sc">+</span></span>
<span id="cb22-5"><a href="#cb22-5"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">colour =</span> <span class="st">"#B30C00"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb22-6"><a href="#cb22-6"></a>  <span class="fu">theme_classic</span>()  <span class="sc">+</span></span>
<span id="cb22-7"><a href="#cb22-7"></a>  <span class="fu">theme</span>(</span>
<span id="cb22-8"><a href="#cb22-8"></a>    <span class="at">panel.grid.major.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb22-9"><a href="#cb22-9"></a>    <span class="at">panel.grid.major.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb22-10"><a href="#cb22-10"></a>    <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">colour =</span> <span class="st">"#8c8c8c"</span>),</span>
<span id="cb22-11"><a href="#cb22-11"></a>    <span class="at">legend.position =</span> <span class="st">"none"</span></span>
<span id="cb22-12"><a href="#cb22-12"></a>  ) <span class="sc">+</span></span>
<span id="cb22-13"><a href="#cb22-13"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste</span>(contrast_name, <span class="st">"MA Plot"</span>, <span class="at">sep =</span> <span class="st">" "</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">

</div>
</section>
<section id="ma-plot-2" class="slide level2 smaller">
<h2>MA Plot</h2>

<img data-src="DESeq2-Visualisation_files/figure-revealjs/unnamed-chunk-28-1.png" width="960" class="r-stretch"><p><strong>These data do not exhibit over dispersion to any great extent</strong></p>
</section>
<section id="overdispersion" class="slide level2 smaller">
<h2>Overdispersion</h2>
<ul>
<li>RNASeq data exhibits overdispersion
<ul>
<li>Variance &gt; Mean</li>
</ul></li>
<li>Model - Negative Binomial</li>
<li>However, does not account for large dispersion observed with low read counts
<ul>
<li>log2 fold changes adjusted, <strong>lfcShrink()</strong></li>
</ul></li>
</ul>
</section>
<section id="overdispersion-1" class="slide level2 smaller">
<h2>Overdispersion</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Example unshrunken results <img data-src="images/maplot_unshrunken.png" style="width:100.0%"></p>
</div><div class="column" style="width:50%;">
<p>Example shrunken results <img data-src="images/maplot_shrunken.png" style="width:100.0%"></p>
</div>
</div>
</section>
<section id="overdispersion-2" class="slide level2 smaller">
<h2>Overdispersion</h2>
<ul>
<li>LfcShrink - shrinkage of the LFC estimates toward zero when the information for a gene is low
<ul>
<li>Low counts</li>
<li>High dispersion values</li>
</ul></li>
</ul>
<p><strong>Shrinking the log2 fold changes will not change the total number of genes that are identified as significantly differentially expressed</strong></p>
</section>
<section id="log-fold-shrinkage" class="slide level2 smaller">
<h2>Log Fold Shrinkage</h2>
<p>Create a results table with the lof fold shrinkage changes</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>lfc_contrast <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds, <span class="at">coef =</span> <span class="fu">resultsNames</span>(dds)[<span class="dv">2</span>],  <span class="at">type =</span> <span class="st">"apeglm"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb23-2"><a href="#cb23-2"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> <span class="co"># Convert to dataframe</span></span>
<span id="cb23-3"><a href="#cb23-3"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"ensembl_gene_id"</span>) <span class="sc">%&gt;%</span> <span class="co"># Create an ensembl gene id column</span></span>
<span id="cb23-4"><a href="#cb23-4"></a>  <span class="fu">left_join</span>(annotations, <span class="at">by =</span> <span class="st">"ensembl_gene_id"</span>)<span class="sc">%&gt;%</span> <span class="co"># merge annotations by ensembl gene id</span></span>
<span id="cb23-5"><a href="#cb23-5"></a>  <span class="fu">left_join</span>(contrast_total_norm_counts, <span class="at">by =</span> <span class="st">"ensembl_gene_id"</span>) <span class="sc">%&gt;%</span> <span class="co"># Merge normalised counts by ensembl gene id</span></span>
<span id="cb23-6"><a href="#cb23-6"></a>  <span class="fu">arrange</span>(padj) <span class="sc">%&gt;%</span> <span class="co"># Sort the dataframe by the p adjusted value ascending</span></span>
<span id="cb23-7"><a href="#cb23-7"></a>  <span class="fu">replace_na</span>(<span class="fu">list</span>(<span class="at">log2FoldChange =</span> <span class="fl">0.001</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="log-fold-shrinkage-1" class="slide level2 smaller">
<h2>Log Fold Shrinkage</h2>
<p>Save the log fold shruken results table to a file</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="co"># Write the results data to file</span></span>
<span id="cb24-2"><a href="#cb24-2"></a><span class="co"># Output file name</span></span>
<span id="cb24-3"><a href="#cb24-3"></a>outfileName <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"lfc"</span>,contrast_name, <span class="st">'genes'</span>, <span class="st">'tsv'</span>, <span class="at">sep =</span> <span class="st">"."</span>)</span>
<span id="cb24-4"><a href="#cb24-4"></a><span class="fu">write_delim</span>(lfc_contrast,</span>
<span id="cb24-5"><a href="#cb24-5"></a>            <span class="at">file =</span> <span class="fu">file.path</span>(resultsFilePath, outfileName, <span class="at">fsep =</span> .Platform<span class="sc">$</span>file.sep),</span>
<span id="cb24-6"><a href="#cb24-6"></a>            <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="log-fold-shrinkage-2" class="slide level2 smaller">
<h2>Log Fold Shrinkage</h2>
<p>Plot the LFC shrunken data as an MA plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>ma_plot <span class="ot">&lt;-</span> lfc_contrast <span class="sc">%&gt;%</span></span>
<span id="cb25-2"><a href="#cb25-2"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(  <span class="at">x =</span> <span class="fu">log10</span>(baseMean), <span class="at">y =</span> log2FoldChange, <span class="at">colour =</span> padj <span class="sc">&lt;</span> <span class="fl">0.1</span>  )) <span class="sc">+</span></span>
<span id="cb25-3"><a href="#cb25-3"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb25-4"><a href="#cb25-4"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">name =</span> <span class="st">'padj &lt; 0.1'</span>, <span class="at">values =</span> <span class="fu">setNames</span>(<span class="fu">c</span>(<span class="st">"#003865"</span>, <span class="st">"#cccccc"</span>), <span class="fu">c</span>(T, F))) <span class="sc">+</span></span>
<span id="cb25-5"><a href="#cb25-5"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="fu">floor</span>(<span class="fu">min</span>(lfc_contrast<span class="sc">$</span>log2FoldChange)), <span class="fu">ceiling</span>(<span class="fu">max</span>(lfc_contrast<span class="sc">$</span>log2FoldChange)), )) <span class="sc">+</span></span>
<span id="cb25-6"><a href="#cb25-6"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">colour =</span> <span class="st">"#B30C00"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb25-7"><a href="#cb25-7"></a>  <span class="fu">theme_classic</span>()  <span class="sc">+</span></span>
<span id="cb25-8"><a href="#cb25-8"></a>  <span class="fu">theme</span>(</span>
<span id="cb25-9"><a href="#cb25-9"></a>    <span class="at">panel.grid.major.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb25-10"><a href="#cb25-10"></a>    <span class="at">panel.grid.major.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb25-11"><a href="#cb25-11"></a>    <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">colour =</span> <span class="st">"#8c8c8c"</span>),</span>
<span id="cb25-12"><a href="#cb25-12"></a>    <span class="at">legend.position =</span> <span class="st">"none"</span></span>
<span id="cb25-13"><a href="#cb25-13"></a>  ) <span class="sc">+</span></span>
<span id="cb25-14"><a href="#cb25-14"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste</span>(contrast_name, <span class="st">"MA Plot"</span>, <span class="at">sep =</span> <span class="st">" "</span>))</span>
<span id="cb25-15"><a href="#cb25-15"></a></span>
<span id="cb25-16"><a href="#cb25-16"></a>ma_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="cell">

</div>
</section>
<section id="log-fold-shrinkage-2-output" class="slide level2 smaller output-location-slide"><h2>Log Fold Shrinkage</h2><div class="cell output-location-slide">
<div class="cell-output-display">
<p><img data-src="DESeq2-Visualisation_files/figure-revealjs/unnamed-chunk-31-1.png" width="960"></p>
</div>
</div></section><section id="volcano-plot" class="slide level2 smaller">
<h2>Volcano Plot</h2>
<p>Add a column to the resuts data frame stating that a gene is significantly either <strong>UP</strong> or <strong>DOWN</strong> regulated. <br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>contrast<span class="ot">&lt;-</span> contrast<span class="sc">%&gt;%</span></span>
<span id="cb26-2"><a href="#cb26-2"></a>  <span class="fu">mutate</span>(<span class="at">Expression =</span> <span class="fu">case_when</span>(log2FoldChange <span class="sc">&gt;=</span> foldchange <span class="sc">&amp;</span>  padj <span class="sc">&lt;=</span> qvalueCutoff <span class="sc">~</span> <span class="st">"Up-regulated"</span>,</span>
<span id="cb26-3"><a href="#cb26-3"></a>                                log2FoldChange <span class="sc">&lt;=</span> <span class="sc">-</span>foldchange <span class="sc">&amp;</span>  padj <span class="sc">&lt;=</span> qvalueCutoff <span class="sc">~</span> <span class="st">"Down-regulated"</span>,</span>
<span id="cb26-4"><a href="#cb26-4"></a>                                <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"Unchanged"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="font-size: 40%;">
<div class="cell">
<div class="cell-output-display">

<table class="table" style="margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;"> ensembl_gene_id </th>
   <th style="text-align:right;"> baseMean </th>
   <th style="text-align:right;"> log2FoldChange </th>
   <th style="text-align:right;"> lfcSE </th>
   <th style="text-align:right;"> stat </th>
   <th style="text-align:right;"> pvalue </th>
   <th style="text-align:right;"> padj </th>
   <th style="text-align:left;"> external_gene_name </th>
   <th style="text-align:left;"> chromosome_name </th>
   <th style="text-align:right;"> start_position </th>
   <th style="text-align:right;"> end_position </th>
   <th style="text-align:left;"> gene_biotype </th>
   <th style="text-align:right;"> entrezgene_id </th>
   <th style="text-align:right;"> Control_1 </th>
   <th style="text-align:right;"> Control_2 </th>
   <th style="text-align:right;"> Control_3 </th>
   <th style="text-align:right;"> Control_4 </th>
   <th style="text-align:right;"> Treatment_1 </th>
   <th style="text-align:right;"> Treatment_2 </th>
   <th style="text-align:right;"> Treatment_3 </th>
   <th style="text-align:right;"> Treatment_4 </th>
   <th style="text-align:left;"> Expression </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> ENSG00000157191 </td>
   <td style="text-align:right;"> 8171.718 </td>
   <td style="text-align:right;"> -2.429669 </td>
   <td style="text-align:right;"> 0.0374863 </td>
   <td style="text-align:right;"> -64.81491 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> NECAP2 </td>
   <td style="text-align:left;"> 1 </td>
   <td style="text-align:right;"> 16440721 </td>
   <td style="text-align:right;"> 16460078 </td>
   <td style="text-align:left;"> protein_coding </td>
   <td style="text-align:right;"> 55707 </td>
   <td style="text-align:right;"> 13556.4812 </td>
   <td style="text-align:right;"> 13858.7451 </td>
   <td style="text-align:right;"> 14103.1250 </td>
   <td style="text-align:right;"> 13621.5255 </td>
   <td style="text-align:right;"> 2555.7914 </td>
   <td style="text-align:right;"> 2514.9013 </td>
   <td style="text-align:right;"> 2576.0675 </td>
   <td style="text-align:right;"> 2587.1079 </td>
   <td style="text-align:left;"> Down-regulated </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000117090 </td>
   <td style="text-align:right;"> 2322.100 </td>
   <td style="text-align:right;"> -5.224265 </td>
   <td style="text-align:right;"> 0.0775600 </td>
   <td style="text-align:right;"> -67.35771 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> SLAMF1 </td>
   <td style="text-align:left;"> 1 </td>
   <td style="text-align:right;"> 160608106 </td>
   <td style="text-align:right;"> 160647044 </td>
   <td style="text-align:left;"> protein_coding </td>
   <td style="text-align:right;"> 6504 </td>
   <td style="text-align:right;"> 4554.9919 </td>
   <td style="text-align:right;"> 4369.2318 </td>
   <td style="text-align:right;"> 4578.5976 </td>
   <td style="text-align:right;"> 4588.9821 </td>
   <td style="text-align:right;"> 118.6919 </td>
   <td style="text-align:right;"> 125.4663 </td>
   <td style="text-align:right;"> 127.4026 </td>
   <td style="text-align:right;"> 113.4383 </td>
   <td style="text-align:left;"> Down-regulated </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000026751 </td>
   <td style="text-align:right;"> 4495.926 </td>
   <td style="text-align:right;"> -4.690935 </td>
   <td style="text-align:right;"> 0.0870932 </td>
   <td style="text-align:right;"> -53.86110 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> SLAMF7 </td>
   <td style="text-align:left;"> 1 </td>
   <td style="text-align:right;"> 160739057 </td>
   <td style="text-align:right;"> 160754821 </td>
   <td style="text-align:left;"> protein_coding </td>
   <td style="text-align:right;"> 57823 </td>
   <td style="text-align:right;"> 8364.7508 </td>
   <td style="text-align:right;"> 8424.9003 </td>
   <td style="text-align:right;"> 9178.0860 </td>
   <td style="text-align:right;"> 8659.3886 </td>
   <td style="text-align:right;"> 366.9206 </td>
   <td style="text-align:right;"> 301.1190 </td>
   <td style="text-align:right;"> 320.1741 </td>
   <td style="text-align:right;"> 352.0674 </td>
   <td style="text-align:left;"> Down-regulated </td>
  </tr>
  <tr>
   <td style="text-align:left;"> ENSG00000072694 </td>
   <td style="text-align:right;"> 2202.988 </td>
   <td style="text-align:right;"> 4.309099 </td>
   <td style="text-align:right;"> 0.1114264 </td>
   <td style="text-align:right;"> 38.67216 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:left;"> FCGR2B </td>
   <td style="text-align:left;"> 1 </td>
   <td style="text-align:right;"> 161663143 </td>
   <td style="text-align:right;"> 161678654 </td>
   <td style="text-align:left;"> protein_coding </td>
   <td style="text-align:right;"> 2213 </td>
   <td style="text-align:right;"> 243.6681 </td>
   <td style="text-align:right;"> 229.0702 </td>
   <td style="text-align:right;"> 186.2775 </td>
   <td style="text-align:right;"> 190.6931 </td>
   <td style="text-align:right;"> 4103.6057 </td>
   <td style="text-align:right;"> 4273.6594 </td>
   <td style="text-align:right;"> 4433.7444 </td>
   <td style="text-align:right;"> 3963.1856 </td>
   <td style="text-align:left;"> Up-regulated </td>
  </tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="volcano-plot-1" class="slide level2 smaller">
<h2>Volcano Plot</h2>
<p>Plot the Log fold change against the log of the adjusted P value</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>vp <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> contrast, <span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange,<span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(padj))) <span class="sc">+</span></span>
<span id="cb27-2"><a href="#cb27-2"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> Expression), <span class="at">size =</span> <span class="dv">3</span><span class="sc">/</span><span class="dv">5</span>)  <span class="sc">+</span></span>
<span id="cb27-3"><a href="#cb27-3"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#003865"</span>, <span class="st">"#CCCCCC"</span>,<span class="st">"#B30C00"</span>)) <span class="sc">+</span></span>
<span id="cb27-4"><a href="#cb27-4"></a>  <span class="fu">xlab</span>(<span class="fu">expression</span>(<span class="st">"Log"</span>[<span class="dv">2</span>]<span class="sc">*</span><span class="st">" Fold Change"</span>)) <span class="sc">+</span></span>
<span id="cb27-5"><a href="#cb27-5"></a>  <span class="fu">ylab</span>(<span class="fu">expression</span>(<span class="st">"-Log"</span>[<span class="dv">10</span>]<span class="sc">*</span><span class="er">~</span> <span class="st">""</span><span class="sc">~</span><span class="fu">italic</span>(P)<span class="sc">~</span><span class="st">"value"</span>)) <span class="sc">+</span></span>
<span id="cb27-6"><a href="#cb27-6"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="sc">-</span>foldchange, foldchange),</span>
<span id="cb27-7"><a href="#cb27-7"></a>             <span class="at">col =</span> <span class="st">"#005C8A"</span>,</span>
<span id="cb27-8"><a href="#cb27-8"></a>             <span class="at">linetype =</span> <span class="st">"dotted"</span>,</span>
<span id="cb27-9"><a href="#cb27-9"></a>             <span class="at">alpha =</span> <span class="fl">0.4</span>,</span>
<span id="cb27-10"><a href="#cb27-10"></a>             <span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb27-11"><a href="#cb27-11"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(qvalueCutoff),</span>
<span id="cb27-12"><a href="#cb27-12"></a>             <span class="at">col =</span> <span class="st">"#005C8A"</span>,</span>
<span id="cb27-13"><a href="#cb27-13"></a>             <span class="at">linetype =</span> <span class="st">"dotted"</span>,</span>
<span id="cb27-14"><a href="#cb27-14"></a>             <span class="at">alpha =</span> <span class="fl">0.4</span>,</span>
<span id="cb27-15"><a href="#cb27-15"></a>             <span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb27-16"><a href="#cb27-16"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb27-17"><a href="#cb27-17"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb27-18"><a href="#cb27-18"></a>        <span class="at">panel.grid.major.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb27-19"><a href="#cb27-19"></a>        <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">colour =</span> <span class="st">"grey80"</span>),</span>
<span id="cb27-20"><a href="#cb27-20"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">30</span>),</span>
<span id="cb27-21"><a href="#cb27-21"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">25</span>),</span>
<span id="cb27-22"><a href="#cb27-22"></a>        <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">25</span>),</span>
<span id="cb27-23"><a href="#cb27-23"></a>        <span class="at">legend.text=</span><span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>)) <span class="sc">+</span></span>
<span id="cb27-24"><a href="#cb27-24"></a>  <span class="fu">guides</span>(<span class="at">colour =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size=</span><span class="fl">1.5</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="volcano-plot-2" class="slide level2 smaller">
<h2>Volcano Plot</h2>

<img data-src="DESeq2-Visualisation_files/figure-revealjs/unnamed-chunk-36-1.png" width="960" class="r-stretch"></section>
<section id="section" class="slide level2">
<h2></h2>
<p>Iterate through the pairwise comparisons write out comma seperated files, MA plots, Volcano plots and heatmaps for each comparison</p>
<div class="cell">

</div>
<div class="footer footer-default">

</div>
</section>

    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="DESeq2-Visualisation_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="DESeq2-Visualisation_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="DESeq2-Visualisation_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="DESeq2-Visualisation_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="DESeq2-Visualisation_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="DESeq2-Visualisation_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="DESeq2-Visualisation_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="DESeq2-Visualisation_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="DESeq2-Visualisation_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="DESeq2-Visualisation_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        target: function(trigger) {
          return trigger.previousElementSibling;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'quarto-reveal',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>