---
title: "DESeq2 Full Analysis"
output: html_notebook
---

# Instal/Load Packages
## CRAN
```{r}
cran.packages <- c("tidyverse",
                   "ggrepel",
                   "RColorBrewer",
                   "pheatmap",
                   "devtools")

cran.load <- function(pkg){
        new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
        if (length(new.pkg)){
          install.packages(new.pkg, dependencies = TRUE)
          }
        sapply(pkg, require, character.only = TRUE)
}
cran.load(cran.packages)
```

## Bioconductor
```{r}
bioconductor.packages <- c("DESeq2",
                           "biomaRt",
                           "apeglm",
                           "org.Hs.eg.db",
                           "clusterProfiler",
                           "enrichplot",
                           "pathview")

bioconductor.load <- function(pkg){
        new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
        if (length(new.pkg)){
          if (!requireNamespace("BiocManager", quietly = TRUE))
            install.packages("BiocManager")
            BiocManager::install(new.pkg)
        }
        sapply(pkg, require, character.only = TRUE)
}
bioconductor.load(bioconductor.packages)
```

## Github
```{r}
git.packages <- c("pipelineTools")
repository <- "GrahamHamilton"
git.load <- function(pkg){
        new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
        if (length(new.pkg))
          install_github(paste(repository,new.pkg,sep = "/"), dependencies = TRUE)
        sapply(pkg, require, character.only = TRUE)
}
git.load(git.packages)
```

## Read in the samples description table
```{r}
sample_table <- read_tsv("./data/sample_description.tsv",col_types = "ff")
head(sample_table)
```

## Read in the comparisons table
```{r}
comparisons <- read_tsv("./data/comparisons.tsv",col_types = "ff")
head(comparisons)
```

## Read in the feature counts table
```{r}
feature_counts_table <- read_tsv("./data/feature_counts_table.tsv") %>%
  column_to_rownames(var = colnames(.[,1])) %>%
  as.data.frame()

# Ensure that the counts table is in the same order as the sample table
feature_counts_table <- feature_counts_table %>%
  dplyr::select(c(sample_table$Sample))

head(feature_counts_table)
```

## Set the cut offs used for filtering results
```{r}
#| echo: true
pvalueCutoff <- 0.05 # P value
qvalueCutoff <- 0.1 # Adjusted P value
foldchange <- 1 # Fold change, usually fold change is log2
```

## Create data directory
```{r}
deseq2_results_dir <- "DESeq2"
dir.create(deseq2_results_dir, showWarnings = FALSE)
```

## Database variable
```{r}
kegg_organism <- "hsa"
annotation_db <- "org.Hs.eg.db"
```

# Differential expression with DESeq2
```{r}
DESeq2Table <- DESeqDataSetFromMatrix(countData = feature_counts_table,
                                      colData = sample_table,
                                      design = ~  Condition)
dds <- DESeq(DESeq2Table)
```

## Save the mormalised read counts
```{r}
total_norm_counts <- estimateSizeFactors(dds) %>%
  counts(normalized = TRUE) %>%
  as.data.frame() %>%
  rownames_to_column(var = "ensembl_gene_id")

write_tsv(total_norm_counts,"normalised_read_counts.tsv")
```


## Dispersion Plot
The black points are the dispersion estimates for each gene as obtained by considering the information from each gene separately.
Unless there are many samples, these values fluctuate around their true values.
The red trend line is fitted, which shows the dispersion dependence on the mean, and then shrink each gene's estimate towards the red line to obtain the final estimates (blue points) that are then used in the hypothesis test.
The blue circles above the main "cloud" of points are genes which have high gene--wise dispersion estimates which are labelled as dispersion outliers.
These estimates are therefore not shrunk toward the fitted trend line.
These warnings just indicate that the dispersion estimation failed for some genes.
```{r}
dispersion_data <- mcols(dds) %>% 
  as.data.frame() %>%
  dplyr::select(baseMean, dispGeneEst, dispFit, dispersion) %>% 
  pivot_longer(-baseMean, names_to = "variable", values_to = "value") %>%
  drop_na()
  
dispersion_plot <- ggplot(dispersion_data, aes(x=baseMean, y=value, colour=variable)) + 
  geom_point(size=0.1) +
  scale_x_log10() + 
  scale_y_log10() + 
  ylab("Dispersion") + 
  xlab("BaseMean") +
  scale_colour_manual(values=c("#999999", "#B30C00", "#003865"),
                      breaks=c("dispGeneEst", "dispFit", "dispersion"),
                      labels=c("Estimate", "Fit", "Final"),
                      name="") +
  theme_classic() + 
  guides(colour = guide_legend(override.aes = list(size=2)))

ggsave(filename = "DispPlot.png",
       plot = dispersion_plot,
       path = deseq2_results_dir)
dispersion_plot
```

## Principle Component Analysis
### Create the matrices of tramnsformed counts
```{r}
rld <- rlog(dds,fitType = "local")
rld_mat <- assay(rld)
vsd <- varianceStabilizingTransformation(dds,fitType = "local")
vsd_mat <- assay(vsd)
```

Calculate the principle component values using the R package prcomp().
Add the meta data from the sample description table.

#### Rlog
```{r}
pca <- prcomp(t(rld_mat)) # Calculate the principle components

pca_data <- pca$x %>%
  as.data.frame() %>% # convert the matrix of PC values to data frame
  rownames_to_column(var = "Sample") %>% # Create a Sample column
  left_join(sample_table) # Add meta data from sample table

percent_var <- round(pca$sdev^2/sum(pca$sdev^2) * 100, digits = 1) # Calculate the percent variance for each PC

# Plot and save the rlog PCA plot
rlog_pca <- ggplot(pca_data, aes(x = PC1, y = PC2, label=Sample, color=Condition)) +
  geom_point(size=5) +
  geom_text_repel(colour = "black",size=3) +
  xlab(paste0("PC1: ",percent_var[1],"% variance")) +
  ylab(paste0("PC2: ",percent_var[2],"% variance")) +
  theme_classic() +
  labs(title = "PCA Plot",
       subtitle = "rlog transformed ") +
  theme(legend.position = "right",
        panel.grid.major.x = element_line("grey"), 
        panel.grid.major.y= element_line("grey"))

ggsave(filename = "rlog_PCA.png",
       plot = rlog_pca,
       path = deseq2_results_dir)
  
rlog_pca
```

#### VSD
```{r}
pca <- prcomp(t(vsd_mat)) # Calculate the principle components

pca_data <- pca$x %>%
  as.data.frame() %>% # convert the matrix of PC values to data frame
  rownames_to_column(var = "Sample") %>% # Create a Sample column
  left_join(sample_table) # Add meta data from sample table

percent_var <- round(pca$sdev^2/sum(pca$sdev^2) * 100, digits = 1) # Calculate the percent variance for each PC

# Plot and save the vsd PCA plot
vsd_pca <- ggplot(pca_data, aes(x = PC1, y = PC2, label=Sample, color=Condition)) +
  geom_point(size=5) +
  geom_text_repel(colour = "black",size=3) +
  xlab(paste0("PC1: ",percent_var[1],"% variance")) +
  ylab(paste0("PC2: ",percent_var[2],"% variance")) +
  theme_classic() +
  labs(title = "PCA Plot",
       subtitle = "vsd transformed ") +
  theme(legend.position = "right",
        panel.grid.major.x = element_line("grey"), 
        panel.grid.major.y= element_line("grey"))

ggsave(filename = "vsd_PCA.png",
       plot = rlog_pca,
       path = deseq2_results_dir)
  
vsd_pca
```

## Heatmap

```{r}
de_genes <- results(dds) %>% # Get the de results
  as.data.frame() %>% # Convert results to a data frame
  rownames_to_column(var = "ensembl_gene_id") %>% # add the ensembl ids as a data frame column
  filter(padj <= qvalueCutoff) %>% # Select data frame rows that have p values below the cut off
  pull(ensembl_gene_id) # get the ensembl ids of the filtered genes as a character vector

log2_de_genes <- total_norm_counts  %>%
  filter(ensembl_gene_id %in% de_genes) %>% # Filter the normalised counts data frame for rows with the ensembl id of the differentially expressed genes list
  column_to_rownames(var = "ensembl_gene_id") %>% # convert the ensembl_gene_id column to the dtaframe rownames
  mutate(across(where(is.numeric), ~ .x + 1)) %>% # Add 1 to all the values to avoid NaN errors
  log2() # Log2 tranform the data frame values

z_scores <- (log2_de_genes - rowMeans(log2_de_genes)) /rowSds(as.matrix(log2_de_genes)) # Calculate z-scores

num_colours = 100 # Number of colour values in gradient

# Set the o value to white
symetric_breaks <- seq(-max(abs(z_scores)),
                       max(abs(z_scores)),
                       length.out = num_colours + 1)

# Set the column annotations
column_annotations <- sample_table %>%
  column_to_rownames(var = "Sample")

# Create the hetamap
ph <- pheatmap(z_scores,
         color =colorRampPalette(c("#003865", "#FFFFFF","#B30C00"))(num_colours),
         breaks = symetric_breaks,
         show_rownames = F,
         annotation_col = column_annotations)

# Save heatmap
save_pheatmap_png(plot = ph, 
                  filename = "Heatmap.png",
                  path = deseq2_results_dir)
```

# Annotations for Biomart
```{r}
database <- "hsapiens_gene_ensembl"
mart <- "genes"
filt <- "ensembl_gene_id"

ensembl <- useEnsembl(biomart = mart, dataset = database)

att <- c("ensembl_gene_id","external_gene_name","chromosome_name","start_position","end_position","gene_biotype","entrezgene_id")

annotations <- getBM(attributes=att,filter=filt,values=rownames(dds),mart=ensembl) %>% 
  distinct(ensembl_gene_id, .keep_all = TRUE)
```

# Pairwise comparisons
```{r}
pairwise_comparisons <- function(comparison){
  base_condition <- comparison[1]
  test_condition <- comparison[2]
  
  # Contrast name
  contrast_name <- paste(test_condition, "vs", base_condition, sep = "_")
  print(contrast_name)
  #############################################################################################
  # Pairwise results
  #############################################################################################
  # Create directory for DESeq2 data
  results_dir = paste(contrast_name, "results", sep = "_")
  resultsFilePath <-file.path(deseq2_results_dir, results_dir, fsep = .Platform$file.sep)
  dir.create(resultsFilePath, showWarnings = FALSE)

  # Get the samples names for the contrast
  comparisonSampleNames <-  sample_table %>%
    filter(Condition %in% c(test_condition, base_condition)) %>%
    pull(Sample)

  # Subset the log2 normailsed counts to just the commpasison samples
  contrast_total_norm_counts <- total_norm_counts %>% 
    column_to_rownames(var = "ensembl_gene_id") %>%
    dplyr::select(all_of(comparisonSampleNames)) %>%
    rownames_to_column(var = "ensembl_gene_id")
  
  # Get the pairwise comparison results
  contrast <- lfcShrink(dds, type = "apeglm", coef = paste("Condition", contrast_name, sep = "_")) %>% # Get the log fold shrunk pair wise comparison results
    as.data.frame() %>% # Convert to data frame
    rownames_to_column(var = "ensembl_gene_id") %>% # Create an ensembl gene id column
    left_join(., annotations, by = "ensembl_gene_id")%>% # merge annotations by ensembl id
    left_join(., contrast_total_norm_counts, by = "ensembl_gene_id") %>% # Merge normalised counts by ensembl id
    arrange(padj) %>% # Sort the data frame by the p adjusted value ascending
    replace_na(list(log2FoldChange = 0.001)) # Replace NA's in logfold change with negligible value


  # Write the results data to file
  # Output file name
  outfileName <- paste(contrast_name, 'genes', 'tsv', sep = ".")
  write.table(contrast,
              file = file.path(resultsFilePath, outfileName, fsep = .Platform$file.sep),
              quote = FALSE,
              row.names = FALSE,
              sep = "\t")
  
  #############################################################################################
  # MA plot
  #############################################################################################
  ma.plot <- contrast %>%
    ggplot(aes(  x = log10(baseMean), y = log2FoldChange, colour = padj < 0.1  )) +
    geom_point(size = 0.5) +
    scale_colour_manual(name = 'padj < 0.1', values = setNames(c("#003865", "#cccccc"), c(T, F))) +
    scale_y_continuous(breaks = seq(floor(min(contrast$log2FoldChange)), ceiling(max(contrast$log2FoldChange)),  1)) +
    geom_hline(yintercept = 0, colour = "#B30C00", alpha = 0.5) +
    theme_classic()  +
    theme(
      panel.grid.major.x = element_blank(),
      panel.grid.major.y = element_blank(),
      axis.line = element_line(colour = "#8c8c8c"),
      legend.position = "none"
    ) +
    labs(title = paste(contrast_name, "MA Plot", sep = " "))
  
  ggsave(filename = paste(contrast_name, 'genes', 'plotMA', 'png', sep = "."),
         path = resultsFilePath,
         plot = ma.plot)
  
  #############################################################################################
  # Heatmap
  #############################################################################################
  de_genes <- contrast %>%
    filter(padj < qvalueCutoff) %>%
    pull(ensembl_gene_id)
  
  if (length(de_genes) > 1) {
    log2_de_genes <- contrast_total_norm_counts %>% # add the ensembl ids as a data frame column
      filter(ensembl_gene_id %in% de_genes) %>% # Filter the normalised counts data frame for rows with the ensembl id of the differentially expressed genes list
      column_to_rownames(var = "ensembl_gene_id") %>% # convert the ensembl_gene_id column to the dtaframe rownames
      mutate(across(where(is.numeric), ~ .x + 1)) %>% # Add 1 to all the values to avoid NaN errors
      log2() # Log2 tranform the data frame values
    
    z_scores <- (log2_de_genes - rowMeans(log2_de_genes)) / rowSds(as.matrix(log2_de_genes)) # Convert to z-scores
  
    num_colours = 100 # Set the number of colours to be created for the heatmapping
    
    # Set the range and mid point, the mid point colour is set to zero
    symetric_breaks <- seq(-max(abs(z_scores)),
                           max(abs(z_scores)),
                           length.out = num_colours + 1)
    
    # Format the annotations file  for pheatmap
    column_annotations <- sample_table %>%
      filter(Sample %in% comparisonSampleNames) %>%
      column_to_rownames(var = "Sample")
    
    # Plot the heatmap
    ph <- pheatmap(z_scores,
             color =colorRampPalette(c("#003865", "#FFFFFF","#B30C00"))(num_colours),
             breaks = symetric_breaks,
             show_rownames = F,
             annotation_col = column_annotations)
    heatmap_name <- paste(contrast_name, qvalueCutoff, "heatmap.png", sep = ".")
    save_pheatmap_png(plot =  ph, 
                      filename = heatmap_name,
                      path = resultsFilePath)
  }
  
  #############################################################################################
  # Volcano plot
  #############################################################################################
  contrast<- contrast%>%
    mutate(Expression = case_when(log2FoldChange >= foldchange &  padj <= qvalueCutoff~ "Up-regulated",
                                  log2FoldChange <= -foldchange &  padj <= qvalueCutoff ~ "Down-regulated",
                                  TRUE ~ "Unchanged"))
  
  vp <- ggplot(data = contrast, aes(x = log2FoldChange,y = -log10(padj))) +
    geom_point(aes(color = Expression), size = 3/5)  +
    scale_color_manual(values = c("#003865", "#CCCCCC","#B30C00")) +
    xlab(expression("Log"[2]*" Fold Change")) +
    ylab(expression("-Log"[10]*~ ""~italic(P)~"value")) +
    geom_vline(xintercept = c(-foldchange, foldchange),
               col = "#005C8A",
               linetype = "dotted",
               alpha = 0.4,
               size = 1) +
    geom_hline(yintercept = -log10(qvalueCutoff),
               col = "#005C8A",
               linetype = "dotted",
               alpha = 0.4,
               size = 1) +
    xlim(-3,2.5) +
    ylim(0, 5) +
    theme_classic() +
    theme(panel.grid.major.x = element_blank(),
          panel.grid.major.y = element_blank(),
          axis.line = element_line(colour = "grey80"),
          plot.title = element_text(size = 30),
          axis.title = element_text(size = 25),
          axis.text = element_text(size = 25),
          legend.text=element_text(size = 15)) +
    guides(colour = guide_legend(override.aes = list(size=1.5)))
  
  ggsave(filename = "volcano_plot.png",
         plot = vp, 
         path = resultsFilePath,
         device = "png")
  
  #############################################################################################
  # GO Analysis
  #############################################################################################
  # Subset the differential expressioxn results to return genes with padj cutoff and 
  filtered_de_genes <- contrast %>%
    filter(abs(log2FoldChange) > foldchange) %>%
    drop_na()
  
  de_entrez_genes <- filtered_de_genes %>%
    pull(entrezgene_id)
  
 # Run the GO enrichment 
  go_ora <- enrichGO(gene = de_entrez_genes,
                     OrgDb = annotation_db,
                     ont = "CC", # Can choose "BP","CC", "MF" or "ALL"
                      pAdjustMethod = "BH", # Can choose "holm", "hochberg", "hommel","bonferroni", "BH","BY", "fdr", "none"
                     pvalueCutoff = pvalueCutoff,
                     qvalueCutoff = qvalueCutoff,
                     readable = TRUE)

  # If the GO analysis produced results, print and save a dotplot and save the results table
  if (nrow(as.data.frame(go_ora)) > 0) {
    ggsave(filename = file.path(resultsFilePath, "GO_dotplot.pdf", fsep = .Platform$file.sep),
           plot = dotplot(go_ora))
    # Write out the results to a file
    go_ora %>%
      as.data.frame() %>%
      write_delim(file = file.path(resultsFilePath, "ora_gene_ontology.tsv"),
                  delim = "\t")
  }
  
  #############################################################################################
  # Pathway Analysis
  #############################################################################################
  
  de_fold_change_list <- filtered_de_genes$log2FoldChange
  names(de_fold_change_list) <- filtered_de_genes$entrezgene_id
  de_fold_change_list <- sort(de_fold_change_list, decreasing = TRUE)
  
  kegg_ora <- enrichKEGG(gene = de_entrez_genes,
                       organism = 'hsa',
                       pvalueCutoff = pvalueCutoff)
  
  # Save the kegg results to file
  kegg_ora %>%
  as.data.frame() %>%
  write_delim(file = file.path(resultsFilePath, "ora_pathway_analysis.tsv"),
              delim = "\t")
  
  # Pathway visualisation function
  kegg_pathview <- function(kegg_pathway_id_list,de_fold_change_list){
  pathview(gene.data = de_fold_change_list, # List of DE gene entrez ids and log fold changes
           pathway.id = kegg_pathway_id_list, # KEGG pathway id to download and annotate
           species = kegg_organism, # Organism KEGG id
           gene.idtype = "KEGG",
           kegg.native = TRUE)
  }
  
  # Get list if differentially expressed KEGG pathways
  kegg_ora_list <- kegg_ora %>%
    as.data.frame() %>%
    pull(ID) 
  
  # Apply the pathway visualisations on the de pathways
  lapply(kegg_ora_list, kegg_pathview, de_fold_change_list = de_fold_change_list)
  
  # Copy the pathway results to the results directory
  results.files <- list.files(path = ".", pattern = "pathview.png") %>% # list all pathview files
    file.copy(resultsFilePath) # Copy pathway files to results directory
  # Delete results files
  remove_files <- list.files(path = ".", pattern = "hsa") %>% # List all files dowloaded from KEGG
    file.remove() # Delete the files
}



apply(comparisons, 1, pairwise_comparisons)
```
